Add support for MFA signup and login flow

michaelschattgen opened the pull request:
## What kind of change does this PR introduce?

This pull request adds support for the multi factor authentication flow. I've tested all the methods through the Example application. Please let me know if I need to add/change anything.

acupofjose commented:
@michaelschattgen this is an awesome PR - thank you for your efforts! Code / implementation all looks solid to me, really appreciate the comments and clarity.

Any chance you'd be willing to include tests too? That's about the only thing I see missing.

michaelschattgen commented:
Thank you! I expect to add tests somewhere in the upcoming days. 😌

michaelschattgen commented:
I've added the tests for both the StatelessClient and the GotrueClient. I've decided to just combine all MFA calls into 1 test since they all overlap, they all rely on a working Enroll, Verify and Challenge. Alongside these complete flow tests I've also added a couple failure tests.

Please let me know if this is sufficient!

acupofjose commented:
Splendid work, seriously. Thanks for being so thorough - Happy to merge this featureset!
